FirebaseDynamicLinks/Sources/FIRDynamicLinkNetworking.m

```diff
@@ -118,33 +119,65 @@ - (void)resolveShortLink:(NSURL *)url
     @"sdk_version" : FDLSDKVersion
   };
 
-  FIRNetworkRequestCompletionHandler resolveLinkCallback = ^(NSData *data, NSError *error) {
-    NSURL *resolvedURL;
-
-    if (!error && data) {
-      NSDictionary *result = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
-      if ([result isKindOfClass:[NSDictionary class]]) {
-        id invitationIDObject = [result objectForKey:@"invitationId"];
-
-        NSString *invitationIDString;
-        if ([invitationIDObject isKindOfClass:[NSDictionary class]]) {
-          NSDictionary *invitationIDDictionary = invitationIDObject;
-          invitationIDString = invitationIDDictionary[@"id"];
-        } else if ([invitationIDObject isKindOfClass:[NSString class]]) {
-          invitationIDString = invitationIDObject;
+  FIRNetworkRequestCompletionHandler resolveLinkCallback = ^(NSData *data, NSURLResponse *response,
+                                                             NSError *error) {
+    NSURL *resolvedURL = nil;
+
+    if (![response isKindOfClass:[NSHTTPURLResponse class]]) {
+      NSError *responseTypeError = [NSError
```

Comment:
The whole method change that I can see is about constructing the error object and passing it through. Can we refactor this with a new method which can take in the NSHTTPURLResponse object and returns an NSError object and move all your code changes there?

Then in this method you can call the new method without have much changes to the existing method. This would make the code more readable and we can test the new method individually. Also avoids duplicated lines of code to call handler(url,error) from multiple if else blocks.

So the new method would be something like (Feel free to change the name and signature of the method) : 

   `+(nullable NSError *) extractErrorFromResponse:(NSURLResponse *)response
                            withError:(nullable NSError *) error{
  if(error){
    return error;
  }

  NSError *customError = nil;

  //Your logic to create the custom error.

  return customError;
}`

In current completion handler, we will just call : 

`error = [FIRDynamicLinkNetworking extractErrorFromResponse:response
             withError:error];`

and the rest of the code will remains same. Let me know if that makes sense.

Reply:
I got the point.

Reply:
Please, see the implementation and if it's ok with you I can add a couple of unit tests.

Reply:
seems good to me apart from a minor comment regarding checking the variable.